### ISSUE TITLE
Errors with 2nd-order derivatives involving convolutions

### ISSUE THREAD

**Issue by levskaya** [google/jax#563]
Justin Gilmer @jmgilmer and I are encountering a few issues with 2nd-order derivatives and convolutions in very simple neural-net settings of chained convolutions and simple nonlinearities.

We've minimized the two bugs we've encountered in pure functions at the public colab repro:
https://colab.research.google.com/drive/1y_Sz3tkJaKgnVgBl8MbigZGtOBGN4nHt

e.g. we can't successfully calculate hessians or HVPs of the below:
```
inputs = lax.conv_general_dilated(inputs, W0, (1,1), 'SAME', (1,1), (1,1), ('NHWC', 'HWIO', 'NHWC'))
inputs = lax.conv_general_dilated(inputs, W1, (2,2), 'SAME', (1,1), (1,1), ('NHWC', 'HWIO', 'NHWC'))
inputs = np.reshape(inputs, (inputs.shape[0], -1))
inputs = np.tanh(inputs)
```

**Comment by mattjj** [comment]
Thanks for raising this and for the minimal repros.

The second issue is a NotImplementedError for a particular batching rule of conv, so that's not a bug so much as something no one has implemented yet.

Not sure about that first one...

**Comment by mattjj** [comment]
This is definitely a bug in our conv transpose rule. The assertions in a0fbff2 catch it earlier.

Here's a smaller repro. I think these inputs are valid, so the bug is in the function being called:

```python
g = np.ones((1, 1, 1, 1), np.float32)
rhs = np.ones((1, 16, 16, 1), np.float32)

params = {
  'lhs_shape': (1, 32, 32, 1),
  'dimension_numbers': lax.ConvDimensionNumbers(
    lhs_spec=(3, 0, 1, 2),
    rhs_spec=(3, 0, 1, 2),
    out_spec=(2, 3, 0, 1)),
  'window_strides': (1, 1),
  'lhs_dilation': (1, 1),
  'padding': ((0, -1), (0, -1)),
  'rhs_shape': (1, 16, 16, 1),
  'rhs_dilation': (2, 2)
}

lax._conv_general_dilated_transpose_lhs(g, rhs, **params)
```

**Comment by mattjj** [comment]
This was fixed in #567.
